x/programs/rust/examples/counter/src/lib.rs

```diff
@@ -59,201 +59,207 @@ pub fn get_value_external(_: Context, target: Program, max_units: i64, of: Addre
+        plan.add_step(Step::create_key(Key::Ed25519(owner_key.clone())));
 
         plan.add_step(Step {
             endpoint: Endpoint::Key,
             method: "key_create".into(),
             params: vec![alice_key.clone()],
             max_units: 0,
-            require: None,
         });
 
         plan.add_step(Step {
             endpoint: Endpoint::Execute,
             method: "program_create".into(),
             max_units: 1000000,
             params: vec![Param::String(PROGRAM_PATH.into())],
-            require: None,
         });
 
-        // run plan
         let plan_responses = simulator.run_plan(plan).unwrap();
 
         assert!(
-            plan_responses.iter().all(|resp| resp.error.is_none()),
+            plan_responses.iter().all(|resp| resp.base.error.is_none()),
             "error: {:?}",
             plan_responses
                 .iter()
-                .filter_map(|resp| resp.error.as_ref())
+                .filter_map(|resp| resp.base.error.as_ref())
                 .next()
         );
     }
 
     #[test]
     fn increment() {
-        let simulator = simulator::Client::new();
+        let mut simulator = simulator::ClientBuilder::new().try_build().unwrap();
 
         let owner_key = String::from("owner");
         let bob_key = Param::Key(Key::Ed25519(String::from("bob")));
 
-        let mut plan = Plan::new(owner_key.clone());
+        let mut plan = Plan::new(&owner_key);
 
-        plan.add_step(Step::create_key(Key::Ed25519(owner_key)));
+        plan.add_step(Step::create_key(Key::Ed25519(owner_key.clone())));
 
         plan.add_step(Step {
             endpoint: Endpoint::Key,
             method: "key_create".into(),
             params: vec![bob_key.clone()],
             max_units: 0,
-            require: None,
         });
 
         let counter_id = plan.add_step(Step {
             endpoint: Endpoint::Execute,
             method: "program_create".into(),
             max_units: 1000000,
             params: vec![Param::String(PROGRAM_PATH.into())],
-            require: None,
         });
 
         plan.add_step(Step {
             endpoint: Endpoint::Execute,
             method: "inc".into(),
             max_units: 1000000,
             params: vec![counter_id.into(), bob_key.clone(), 10.into()],
-            require: None,
         });
 
-        plan.add_step(Step {
-            endpoint: Endpoint::ReadOnly,
-            method: "get_value".into(),
-            max_units: 0,
-            params: vec![counter_id.into(), bob_key.clone()],
-            require: Some(Require {
-                result: ResultAssertion::NumericEq(10),
-            }),
-        });
-
-        // run plan
         let plan_responses = simulator.run_plan(plan).unwrap();
 
-        // ensure no errors
         assert!(
-            plan_responses.iter().all(|resp| resp.error.is_none()),
+            plan_responses.iter().all(|resp| resp.base.error.is_none()),
```

Comment:
Yeah, I don't like this. Using a custom serializer will be better. I think I said it in my other comment, but you can do this as a follow-up.

Reply:
I don't like it either, and I don't want this to get merged so I will work on the custom serializer

Reply:
Me neither, but this is a temporary solution.
Once we drop `Plan`, all of the `PlanResponse` will be typed and the intermediary types will be removed, directly inlined in the `PlanResponse` struct, with a custom `Serializer`.